Add IMDSv2 support for AWS EBS builder

hubbcaps opened the pull request:
This MR adds the ability to set three new config options for the AWS EBS builder:

- http_endpoint
- http_tokens
- http_put_response_hop_limit

These options allow us to set the metadata options to be able to utilize the new IMDSv2 service inside AWS, which is required for environments that are starting to crop up where IMDSv2 is enforced account wide in the wake of the Capital One hack that was precipitated on IMDSv1.

Notes, this was my first foray into Go. Any advice or recommendations is appreciated, I wasn't quite sure if my implementation in step_run_source_instance.go and step_run_spot_instance.go for talking to the AWS SDK was correct, it looked a bit long in combination with everything else around it and I'm wondering if there isn't a better way to combine the three metadata options into a single option to be passed there.

Closes #9338

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/packer?pullRequest=10546)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/packer?pullRequest=10546)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/packer?pullRequest=10546).</sub>

sylviamoss commented on builder/amazon/common/run_config.go:
```diff
@@ -426,6 +426,18 @@ type RunConfig struct {
 	// 10m
 	WindowsPasswordTimeout time.Duration `mapstructure:"windows_password_timeout" required:"false"`
 
+	// Metadata Options
+	// See [Configure IMDS](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html) for details.
+	// A string to enable or disble the IMDS endpoint for an instance. Defaults to enabled.
+	// Accepts either "enabled" or "disabled"
+	HttpEndpoint string `mapstructure:"http_endpoint" required:"false"`
+	// A string to either set the use of IMDSv2 for the instance to optional or required. Defaults to "optional".
+	// Accepts either "optional" or "required"
+	HttpTokens string `mapstructure:"http_tokens" required:"false"`
+	// A numerical value to set an upper limit for the amount of hops allowed when communicating with IMDS endpoints.
+	// Defaults to 1.
+	HttpPutResponseHopLimit int64 `mapstructure:"http_put_response_hop_limit" required:"false"`
```
I suggest exporting this into it's own struct like 

```
// Metadata Options
// See [Configure IMDS](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html) for details.
type MetadataOptions struct {
	// A string to enable or disble the IMDS endpoint for an instance. Defaults to enabled.
	// Accepts either "enabled" or "disabled"
	HttpEndpoint string `mapstructure:"http_endpoint" required:"false"`
	// A string to either set the use of IMDSv2 for the instance to optional or required. Defaults to "optional".
	// Accepts either "optional" or "required"
	HttpTokens string `mapstructure:"http_tokens" required:"false"`
	// A numerical value to set an upper limit for the amount of hops allowed when communicating with IMDS endpoints.
	// Defaults to 1.
	HttpPutResponseHopLimit int64 `mapstructure:"http_put_response_hop_limit" required:"false"`
}
```

Then in the RunConfig will be like 
```suggestion
Metadata MetadataOptions `mapstructure:"metadata_options" required:"false"`
```

This way you can use the `metadata_options` in the hcl2 config like: 
```
  metadata_options {
    http_tokens = "required"
    http_put_response_hop_limit = 64
  }
```

sylviamoss replied:
Doing this, at the very top of the file you will have to add the MetadataOptions struct to the `//go:generate mapstructure-to-hcl2` list

hubbcaps replied:
I've broken this into the struct style requested. Let me know if that looks like you expected.

sylviamoss commented on builder/amazon/common/step_run_source_instance.go:
```diff
@@ -114,6 +117,7 @@ func (s *StepRunSourceInstance) Run(ctx context.Context, state multistep.StateBa
 		UserData:            &userData,
 		MaxCount:            aws.Int64(1),
 		MinCount:            aws.Int64(1),
+		MetadataOptions:     &ec2.InstanceMetadataOptionsRequest{HttpEndpoint: &s.HttpEndpoint, HttpTokens: &s.HttpTokens, HttpPutResponseHopLimit: &s.HttpPutResponseHopLimit},
```
I think some validation is necessary before this. We need to make sure `HttpEndpoint` is `enabled` before setting the other two variables. 
ref: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html

I suggest you look at the terraform provider logic to implement something similar 
https://github.com/hashicorp/terraform-provider-aws/blob/main/aws/resource_aws_instance.go#L1364

hubbcaps replied:
I've done some validation for this, it required me to do some sane default setting to make the validation work as expected without flopping back to weird values. Let me know if my implementation is ok.

sylviamoss commented on builder/amazon/ebs/builder.go:
```diff
@@ -187,6 +187,9 @@ func (b *Builder) Run(ctx context.Context, ui packersdk.Ui, hook packersdk.Hook)
 			Debug:                             b.config.PackerDebug,
 			EbsOptimized:                      b.config.EbsOptimized,
 			ExpectedRootDevice:                "ebs",
+			HttpEndpoint:                      b.config.HttpEndpoint,
```
These changes should be added to the other ebs builders as well. The steps and `RunConfig` are shared between them.

hubbcaps replied:
Propagated these additions to the other builders.

sylviamoss commented on builder/amazon/common/run_config.go:
```diff
@@ -426,6 +426,18 @@ type RunConfig struct {
+	// A numerical value to set an upper limit for the amount of hops allowed when communicating with IMDS endpoints.
+	// Defaults to 1.
+	HttpPutResponseHopLimit int64 `mapstructure:"http_put_response_hop_limit" required:"false"`
+
```
You should definitely add some validation in `(c *RunConfig) Prepare(ctx *interpolate.Context) []error` to verify if the options are set with valid values. Like: 

`http_endpoint` should be either "disabled" or "enabled"
`http_put_response_hop_limit` should be between 1 and 64
`http_tokens` should be either "optional" or "required"

hubbcaps replied:
Setup the validation. Note that with the `http_put_response_hop_limit` one, I couldn't get the check to match 0 specifically, so I did a check of 0<64 and if they don't set the option, the default instantiation of it should be 0, so my sane default checks should catch the 0 in this case and flip it to the default of a 1. Let me know if this is acceptable, or you can see why my != 0 check wasn't working, as I tried it a few different ways and worked out the true/false logic on a piece of paper to make sure I wasn't just being stupid with my assertions.

sylviamoss replied:
I suggested some changes to the validation where the defaults are set. Could you see if that solves the problem? I think it will.

sylviamoss reviewed (changes requested):
This is looking good! There's still some work to do and added some comments about it. Let me know if you need any help with it.

hubbcaps commented:
@sylviamoss this is extremely helpful. Thank you for the breakdowns and all the information, I'll jump on this once I have some free time later today and get back to each of your discussion threads as I can.

hubbcaps commented:
Also, if you could help me out and point out what I'm doing wrong with the go.sum, it seems to always have differences compared to the local files I pull down when just running make, not sure why that is and don't wanna clutter the MR with commits fixing it manually if there's something I could do better.

sylviamoss commented on builder/amazon/common/run_config.go:
```diff
@@ -486,6 +503,26 @@ func (c *RunConfig) Prepare(ctx *interpolate.Context) []error {
 	// Validation
 	errs := c.Comm.Prepare(ctx)
 
+	if c.Metadata.HttpEndpoint != "enabled" &&
+		c.Metadata.HttpEndpoint != "disabled" &&
+		c.Metadata.HttpEndpoint != "" {
+		msg := fmt.Errorf("http_endpoint requires either disabled or enabled as its value")
+		errs = append(errs, msg)
+	}
+
+	if c.Metadata.HttpTokens != "optional" &&
+		c.Metadata.HttpTokens != "required" &&
+		c.Metadata.HttpTokens != "" {
+		msg := fmt.Errorf("http_tokens requires either optional or required as its value")
+		errs = append(errs, msg)
+	}
+
+	if c.Metadata.HttpPutResponseHopLimit < 0 ||
+		c.Metadata.HttpPutResponseHopLimit > 64 {
+		msg := fmt.Errorf("http_put_response_hop_limit requires a number between 1 and 64")
+		errs = append(errs, msg)
+	}
```
This is looking good! I would change a bit like:
```suggestion
      if c.Metadata.HttpEndpoint == "" {
		c.Metadata.HttpEndpoint = "enabled"
	}
	if c.Metadata.HttpTokens == "" {
		c.Metadata.HttpTokens = "optional"
	}
	if c.Metadata.HttpPutResponseHopLimit == 0 {
		c.Metadata.HttpPutResponseHopLimit = 1
	}

	if c.Metadata.HttpEndpoint != "enabled" && c.Metadata.HttpEndpoint != "disabled" {
		msg := fmt.Errorf("http_endpoint requires either disabled or enabled as its value")
		errs = append(errs, msg)
	}

	if c.Metadata.HttpTokens != "optional" && c.Metadata.HttpTokens != "required" {
		msg := fmt.Errorf("http_tokens requires either optional or required as its value")
		errs = append(errs, msg)
	}

	if c.Metadata.HttpPutResponseHopLimit < 0 ||
		c.Metadata.HttpPutResponseHopLimit > 64 {
		msg := fmt.Errorf("http_put_response_hop_limit requires a number between 1 and 64")
		errs = append(errs, msg)
	}
```
This adds the default values and whenever something is set different from the requirements it will fail the validation.

sylviamoss commented on builder/amazon/common/step_run_source_instance.go:
```diff
@@ -144,6 +147,23 @@ func (s *StepRunSourceInstance) Run(ctx context.Context, state multistep.StateBa
 		runOpts.CreditSpecification = &ec2.CreditSpecificationRequest{CpuCredits: &creditOption}
 	}
 
+	// Set Metadata defaults if not defined at all
+	if s.HttpEndpoint == "" {
+		s.HttpEndpoint = "enabled"
+	}
+
+	if s.HttpTokens == "" {
+		s.HttpTokens = "optional"
+	}
+
+	if s.HttpPutResponseHopLimit == 0 {
+		s.HttpPutResponseHopLimit = 1
+	}
```
You don't need this here anymore cuz it should be done on `Prepare`. This way you don't need to set the defaults everywhere you use it. 
```suggestion
```

sylviamoss commented on builder/amazon/common/step_run_spot_instance.go:
```diff
@@ -127,6 +130,23 @@ func (s *StepRunSpotInstance) CreateTemplateData(userData *string, az string,
 
 	}
 
+	// Set Metadata defaults if not defined at all
+	if s.HttpEndpoint == "" {
+		s.HttpEndpoint = "enabled"
+	}
+
+	if s.HttpTokens == "" {
+		s.HttpTokens = "optional"
+	}
+
+	if s.HttpPutResponseHopLimit == 0 {
+		s.HttpPutResponseHopLimit = 1
+	}
```
Same as the above
```suggestion
```

sylviamoss commented on website/content/partials/builder/amazon/common/MetadataOptions-not-required.mdx:
```diff
@@ -0,0 +1,10 @@
+<!-- Code generated from the comments of the MetadataOptions struct in builder/amazon/common/run_config.go; DO NOT EDIT MANUALLY -->
```
You should these two newly generated docs files to the builders' documentation otherwise this won't be shown anywhere in the docs. 

All the amazon builders docs are https://github.com/hashicorp/packer/tree/master/website/content/docs/builders/amazon

sylviamoss commented on website/content/partials/builder/amazon/common/RunConfig-not-required.mdx:
```diff
@@ -364,6 +364,8 @@
   password for Windows instances. Defaults to 20 minutes. Example value:
   10m
 
+- `metadata_options` (MetadataOptions) - Metadata Settings
```
You can add a new section in the builders' documentation to show the config for this one. 

Like: 
```
#### Metadata Settings

@include 'builder/amazon/common/MetadataOptions.mdx'

@include 'builder/amazon/common/MetadataOptions-not-required.mdx'

// Some example 

```
I suggest adding after the `### Run Configuration` section

sylviamoss commented on builder/amazon/common/run_config.go:
```diff
@@ -426,6 +440,9 @@ type RunConfig struct {
 	// 10m
 	WindowsPasswordTimeout time.Duration `mapstructure:"windows_password_timeout" required:"false"`
 
+	// Metadata Settings
```
Once you updated the changes, you can make this line like:
```suggestion
	// [Metadata Settings](#metadata-settings) 
```
This will create a link to the Metadata Settings section.

sylviamoss reviewed (commented):
Thanks for going over my suggestions, this almost done! I added some other suggestions and some will help you with the validation problem. 

I haven't tested the code I suggested, so feel free to change it if something is not right.

sylviamoss commented:
Try rebasing your branch with master and see if that fixes the go.mod problem

hubbcaps commented:
@sylviamoss I've updated based on your recommendations.

However, I did the rebase based on your suggestion, however it ended up causing me some weird mod issues.
I think it has to do with this line added to go.mod 3 days ago by @azr but I'm not really sure
https://github.com/hashicorp/packer/blob/540effbbc0da42e5ca3abe4af277d4d96f0702c4/go.mod#L103

Doing anything with go just kept prompting
```[packer] go mod graph                                                                                                                                                                                                                                                                                                 master 
go: github.com/hashicorp/packer-plugin-sdk@v0.0.11 requires
	github.com/hashicorp/packer@v1.6.7-0.20210126105722-aef4ced967ec: invalid version: unknown revision aef4ced967ec```
	
I tried out what the comment suggested as far as pointing it to a local path after pulling it with go get github.com/hashicorp/packer-plugin-sdk but that didn't seem to work. Eventually I just fell back to a working go.mod and go.sum from my earlier commits to get around it and run my make fmt and make generate to clean up the errors left after working on your suggestions. I updated my commit after running those to again include the latest go.mod and go.sum, but definitely doesn't seem like the ideal practice lol

I'm sure my naivety about wider go development lead to some of these issues and if I was more experienced probably figured out what I was actually doing wrong rather than ham fisting my solution.

Now it seems like a slew of extra files have been edited and I have no idea why that occurred. Please let me know if you have a solution to clean the PR back up, or if I should just start over with a new MR with the cleaned up code.

SwampDragons commented:
I'll push a fix to remove that erroneous line from the modfile now.

SwampDragons commented:
Try copying the go.mod file from this PR: https://github.com/hashicorp/packer/pull/10578 and rebuilding; hopefully that'll sort your PR out and get it back in line with master

hubbcaps commented:
@SwampDragons Using the latest go.mod from that PR I just keep getting the same error I was before and not quite sure how to fix it.

```
[packer] make                                                                                                                                                                                  add-imdsv2-support  ✭
go: github.com/hashicorp/packer-plugin-sdk@v0.0.11 requires
	github.com/hashicorp/packer@v1.6.7-0.20210126105722-aef4ced967ec: invalid version: unknown revision aef4ced967ec
make: *** [Makefile:52: install-build-deps] Error 1
[packer] go mod tidy                                                                                                                                                                           add-imdsv2-support  ✭
go: github.com/hashicorp/packer-plugin-sdk@v0.0.11 requires
	github.com/hashicorp/packer@v1.6.7-0.20210126105722-aef4ced967ec: invalid version: unknown revision aef4ced967ec
[packer] go mod download                                                                                                                                                                       add-imdsv2-support  ✭
go: github.com/hashicorp/packer-plugin-sdk@v0.0.11 requires
	github.com/hashicorp/packer@v1.6.7-0.20210126105722-aef4ced967ec: invalid version: unknown revision aef4ced967ec
```

Also not too sure on how to fix the 451 file additions, unless that's intentional from rebuilding the docs?

sylviamoss commented:
I merged the packer master branch into yours to show the real diff here.

sylviamoss commented on website/content/docs/builders/amazon/ebs.mdx:
````diff
@@ -91,6 +91,64 @@ necessary for this build to succeed and can be found further down the page.
+  "builders": [
+    {
+      "type": "amazon-ebs",
+      "access_key": "{{user `aws_access_key`}}",
+      "secret_key": "{{user `aws_secret_key`}}",
+      "region": "us-east-1",
+      "source_ami": "ami-fce3c696",
+      "instance_type": "t2.micro",
+      "ssh_username": "ubuntu",
+      "ami_name": "packer_AWS {{timestamp}}"
+      "metadata_options": {
+        "http_endpoint": "enabled",
+        "http_tokens": "required",
+        "http_put_response_hop_limit": 1
+      }
+    }
+  ]
+}
+```
+
+</Tab>
+<Tab heading="HCL2">
+
+```hcl
+source "amazon-ebs" "basic-example" {
+  region        =  "us-east-1"
+  source_ami    =  "ami-fce3c696"
+  instance_type =  "t2.micro"
+  ssh_username  =  "ubuntu"
+  ami_name      =  "packer_AWS_example_{{timestamp}}"
+  metadata_options {
+    http_endpoint = enabled
+    http_tokens = required
+    http_put_response_hop_limit = 1
+  }
+}
+```
+
+</Tab>
+</Tabs>
````
Let's just switch the order of the tabs here because we are looking to prioritize the HCL2 from now on.

sylviamoss commented on website/content/docs/builders/amazon/ebs.mdx:
```diff
@@ -91,6 +91,64 @@ necessary for this build to succeed and can be found further down the page.
 
 @include 'builder/amazon/common/RunConfig-not-required.mdx'
 
+#### Metadata Settings
```
Nice!! Thanks for making the changes :D 
Don't forget you must add these to the other ebs builder docs as well.

sylviamoss commented on builder/amazon/common/run_config.go:
```diff
@@ -44,6 +44,20 @@ type SecurityGroupFilterOptions struct {
 	config.NameValueFilter `mapstructure:",squash"`
 }
 
+// [Metadata Options](#metadata-settings)
```
This link points to the current section. Check it out: https://packer-og00506pb.vercel.app/docs/builders/amazon/ebs#metadata-settings
```suggestion
// Configures the metadata options.
```

sylviamoss commented on website/content/docs/builders/amazon/ebs.mdx:
```diff
@@ -91,6 +91,64 @@ necessary for this build to succeed and can be found further down the page.
+      "source_ami": "ami-fce3c696",
+      "instance_type": "t2.micro",
+      "ssh_username": "ubuntu",
+      "ami_name": "packer_AWS {{timestamp}}"
```
Missing `comma`
```suggestion
      "ami_name": "packer_AWS {{timestamp}}",
```

sylviamoss commented on website/content/docs/builders/amazon/ebs.mdx:
```diff
@@ -91,6 +91,64 @@ necessary for this build to succeed and can be found further down the page.
+  ssh_username  =  "ubuntu"
+  ami_name      =  "packer_AWS_example_{{timestamp}}"
+  metadata_options {
+    http_endpoint = enabled
+    http_tokens = required
```
```suggestion
    http_endpoint = "enabled"
    http_tokens = "required"
```

sylviamoss reviewed (approved):
I think the go code is looking good! I tested it with an existing template I have and it works. 👍🏼 

There's still some documentation changes to be made but nothing complex.

hubbcaps commented:
@sylviamoss pushed those requested changes.

Unfortunately that error above about the packer SDK is still haunting me, so I can't run go/make generate or go/make fmt still, or make in general.

sylviamoss commented:
We are currently in a work in progress with the sdk so branches get outdated fast and this has been causing some errors when trying to run/build/generate and etc. I ran the generate for you and updated your branch.

Thanks for working hard on this one! 😄

hubbcaps commented:
Absolutely @sylviamoss, it was a ton of fun. Thanks for all your guidance and insight, it was really helpful.

KevinHock commented:
Awwww yeah, thanks all ;)
